SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestNetReference/SpatialTestNetReference.cpp

```diff
@@ -0,0 +1,136 @@
+
+		// Move the character to the correct location
+		if(Test->TimerHelper == 0.0f)
+		{
+			PlayerCharacter->ServerMoveToLocation(Test->TestLocations[CurrentMoveIndex].Key);
+		}
+
+		// After arriving at the correct location, perform the required checks
+		if (PlayerCharacter->GetActorLocation().Equals(Test->TestLocations[CurrentMoveIndex].Key, 0.25f))
+		{
+			// Wait for a second as it may take some time for the references to be updated correctly
+			if (Test->TimerHelper < 1.0f)
+			{
+				Test->TimerHelper += DeltaTime;
+				return;
+			}
+
+			// Reset the timer so that it can be used for the next move
+			Test->TimerHelper = 0.0f;
+
+			TArray<AActor*> CubesWithReferences;
+			UGameplayStatics::GetAllActorsOfClass(Test->GetWorld(), ACubeWithReferences::StaticClass(), CubesWithReferences);
+
+			// Check the number of relevant cubes
+			Test->AssertEqual_Int(CubesWithReferences.Num(), Test->TestLocations[CurrentMoveIndex].Value, TEXT("Correct number of references are relevant to the player"));
+
+			for (AActor* ArrayObject : CubesWithReferences)
+			{
+				ACubeWithReferences* CurrentCube = Cast<ACubeWithReferences>(ArrayObject);
+				FVector CurrentCubeLocation = CurrentCube->GetActorLocation();
+
+				bool bHasCorrectReferences = true;
+				int ValidReferences = 0;
+
+				for (AActor* OtherObject : CubesWithReferences)
+				{
+					ACubeWithReferences* OtherCube = Cast<ACubeWithReferences>(OtherObject);
+					FVector OtherCubeLocation = OtherObject->GetActorLocation();
+
+					// If the cube is the current one or the diagonally opposed one, then ignore it as it should never be a neigbhour of the current cube
+					if(OtherCubeLocation.Equals(CurrentCubeLocation) || (OtherCubeLocation.X == -CurrentCubeLocation.X && OtherCubeLocation.Y == -CurrentCubeLocation.Y))
+					{
+						continue;
+					}
+
+					// Check that the current cube has a neighbour reference to this OtherCube
+					bHasCorrectReferences &= (CurrentCube->Neighbour1 == OtherCube) || (CurrentCube->Neighbour2 == OtherCube);
+
+					if (bHasCorrectReferences)
+					{ 
+						++ValidReferences;
+					}
+				}
+
+				if (ValidReferences == 0)
+				{
+					// Check that the current cube has 0 valid references
+					bHasCorrectReferences &= !IsValid(CurrentCube->Neighbour1) && !IsValid(CurrentCube->Neighbour2);
+				}
+				else if (ValidReferences == 1)
+				{
+					// We have previously checked that one neighbour reference is correctly pointing to the neighbour cube, also check that the other reference is null
+					bHasCorrectReferences &=  !IsValid(CurrentCube->Neighbour1) || !IsValid(CurrentCube->Neighbour2);
+				}
+
+				Test->AssertEqual_Bool(bHasCorrectReferences, true, TEXT("References are correct"));
+			}
+
+			// Make the character move to the next step in the sequence
+			Test->LocationIndex++;
+
+			// After checking all locations twice, finish the test
+			if (Test->LocationIndex == 2 * Test->TestLocations.Num())
```

Comment:
Did you check breaking some stuff makes the test fails ? 
Hint : you would have to look at USpatialActorChannel::ObjectReferenceMap since this supports unmapping and remapping references.
Try to have the test fail at various stages, like succeed after the first pass, but break at the second.

Reply:
I see, I've managed to make the entire test break by making all stages fail, however, I did not think about trying to make it break after the first pass. Will do that!